NanoProd/plugins/DiLeptonBuilder.cc

```diff
@@ -218,22 +242,48 @@ void DiLeptonBuilder<Lepton1, Lepton2>::produce(edm::StreamID, edm::Event& evt,
   for(size_t l1_idx = 0; l1_idx < leptons1->size(); ++l1_idx) {
     const Lepton1& l1 = leptons1->at(l1_idx);
     if(l1_selection_ && !(*l1_selection_)(l1)) continue;
-    if(src1Veto_ && dR_match(l1, *vetoLeptons1, deltaR_thr_)) continue;
+
+    // Get track associated to l1
+    reco::Track track_l1;
+    if (useStandalone_l1_) {
+      // Accept only pat::Muon objects with an outer track
+      if (!(outerTrackAvailable)(l1)) continue; 
+      track_l1 = getOuterTrack(l1);
+    } else {
+    track_l1 = getTrack(l1);
+    }
+    // Apply veto on l1 using deltaR matching with chosen l1 track 
+    if(src1Veto_ && dR_match(track_l1, *vetoLeptons1, deltaR_thr_)) continue;
 
     const size_t l2_start = l1l2_interchangeable_ && commonSrc_ ? l1_idx + 1 : 0;
     for(size_t l2_idx = l2_start; l2_idx < leptons2->size(); ++l2_idx) {
       const Lepton2& l2 = leptons2->at(l2_idx);
       if(Helper::isSameObject(l1, l2)) continue;
       if(l2_selection_ && !(*l2_selection_)(l2)) continue;
-      if(src2Veto_ && dR_match(l2, *vetoLeptons2, deltaR_thr_)) continue;
+
+      // Get track associated to l2
+      reco::Track track_l2;
+      if (useStandalone_l2_) {
+        if (!(outerTrackAvailable)(l2)) continue;
+        track_l2 = getOuterTrack(l2);
+      } else {
+        track_l2 = getTrack(l2);
+      }
+      // Apply veto on l2 using deltaR matching with chosen l2 track 
+      if(src2Veto_ && dR_match(track_l2, *vetoLeptons2, deltaR_thr_)) continue;
 
       pat::CompositeCandidate lepton_pair;
-      lepton_pair.setP4(getP4(l1) + getP4(l2));
-      lepton_pair.setCharge(l1.charge() + l2.charge());
-      lepton_pair.addUserFloat("lep_deltaR", reco::deltaR(l1, l2));
+      /* Can't always use track here because muon mass is assumed in P4 computation of reco::Track */
+      if (useStandalone_l1_ && useStandalone_l2_) {lepton_pair.setP4(getP4(track_l1) + getP4(track_l2));}
+      else if (useStandalone_l1_) {lepton_pair.setP4(getP4(track_l1) + getP4(l2));}
+      else if (useStandalone_l2_) {lepton_pair.setP4(getP4(l1) + getP4(track_l2));}
+      else {lepton_pair.setP4(getP4(l1) + getP4(l2));}
+
+      lepton_pair.setCharge(track_l1.charge() + track_l2.charge());
+      lepton_pair.addUserFloat("lep_deltaR", reco::deltaR(track_l1, track_l2));
 
       // Put the lepton passing the corresponding selection
-      if(!l1l2_interchangeable_ || l1.pt() >= l2.pt()) {
+      if(!l1l2_interchangeable_ || track_l1.pt() >= track_l2.pt()) {
```

Comment:
Just to note that using the pT from the track here isn't what you'd for electrons, but a) we also don't foresee a two-electron analysis and b) here it's just for the sorting so it won't make any results wrong so this seems fine to me as is.